File: library/src/main/java/org/meganekkovr/Entity.java
package org.meganekkovr;

import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.ArrayMap;
import android.view.View;

import org.joml.Matrix4f;
import org.joml.Quaternionf;
import org.joml.Vector3f;
import org.meganekkovr.animation.EntityAnimator;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Entity is a basic class in Meganekko. It is something on VR scene.
 * Entity has transform information such as {@link #setPosition(float, float, float) position},
 * {@link #setScale(float, float, float) scale}, and {@link #setRotation(Quaternionf) rotation}.
 * It can have child Entities and {@link Component}s.
 * Entity which will be rendered has to have {@link GeometryComponent} and {@link SurfaceRendererComponent}.
 */
public class Entity {
    private final NativePointer nativePointer;
    private final Map<Class<? extends Component>, Component> components = new ArrayMap<>();
    private final List<Entity> children = new CopyOnWriteArrayList<>();
    private final Vector3f position = new Vector3f();
    private final Vector3f scale = new Vector3f(1, 1, 1);
    private final Quaternionf rotation = new Quaternionf();
    private final Matrix4f localMatrix = new Matrix4f();
    private final Matrix4f worldModelMatrix = new Matrix4f();
    private final float[] matrixValues = new float[16];
    private MeganekkoApp app;
    private Entity parent;
    private boolean localMatrixUpdateRequired = true;
    private boolean worldMatrixUpdateRequired = true;
    private int id;
    private float opacity = 1.0f;
    private boolean updateOpacityRequired;
    private boolean visible = true;
    private boolean renderable; // true if this has Geometry and Surface.

    public Entity() {
        nativePointer = NativePointer.getInstance(newInstance());
    }

    private static native void setWorldModelMatrix(long nativePtr, float[] matrix);

    /**
     * Create Entity from {@link View}. New Entity has plane geometry.
     *
     * @param view View for surface.
     * @return new Entity
     */
    @NonNull
    public static Entity from(@NonNull View view) {

        final Entity entity = new Entity();
        entity.add(SurfaceRendererComponent.from(view));
        entity.add(GeometryComponent.from(view));

        return entity;
    }

    /**
     * Create Entity from {@link Drawable}. New Entity has plane geometry.
     *
     * @param drawable Drawable for surface.
     * @return new Entity
     */
    @NonNull
    public static Entity from(@NonNull Drawable drawable) {

        final Entity entity = new Entity();
        entity.add(SurfaceRendererComponent.from(drawable));
        entity.add(GeometryComponent.from(drawable));

        return entity;
    }

    /**
     * Override this to create own native instance.
     * Native class must be derived from {@code mgn::Entity}.
     *
     * @return Native pointer value return from C++ {@code new}.
     */
    protected native long newInstance();

    /**
     * This method is not valid until this is attached to {@link Scene}.
     *
     * @return MeganekkoApp.
     */
    public MeganekkoApp getApp() {
        return app;
    }

    /**
     * For internal use only.
     *
     * @param app
     */
    void setApp(@NonNull MeganekkoApp app) {
        this.app = app;

        // Propagate to children
        for (Entity child : children) {
            child.setApp(app);
        }
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    /**
     * Set string id.
     *
     * @param id ID
     */
    public void setId(@NonNull String id) {
        setId(id.hashCode());
    }

    /**
     * Find {@link Entity} from children. Typically, id is {@code R.id.xxx} value.
     *
     * @param id ID
     * @return Found Entity or {@code null} if it has no matched Entity with id.
     */
    @Nullable
    public Entity findById(int id) {

        if (this.id == id) return this;

        for (Entity child : children) {
            Entity found = child.findById(id);
            if (found != null) return found;
        }

        return null;
    }

    /**
     * Find {@link Entity} from children.
     *
     * @param id ID
     * @return Found Entity or {@code null} if it has no matched Entity with id.
     */
    public Entity findById(@NonNull String id) {
        return findById(id.hashCode());
    }

    /**
     * Called at every frame update.
     *
     * @param frame Frame information
     */
    public void update(@NonNull FrameInput frame) {

        // Notify to components
        for (Component component : components.values()) {
            component.update(frame);
        }

        // Notify to children
        for (Entity child : children) {
            child.update(frame);
        }

        // Update local model matrix if necessary.
        if (localMatrixUpdateRequired) {
            updateLocalMatrix();
            invalidateWorldModelMatrix();
            localMatrixUpdateRequired = false;
        }

        // Update world model matrix if necessary.
        if (worldMatrixUpdateRequired) {
            updateWorldModelMatrix();
            worldMatrixUpdateRequired = false;

            // Update native side values
            worldModelMatrix.get(matrixValues);
            setWorldModelMatrix(nativePointer.get(), matrixValues);
        }

        // Update opacity if necessary.
        if (updateOpacityRequired) {
            updateOpacity();
            updateOpacityRequired = false;
        }
    }

    /**
     * Update local matrix.
     */
    public void updateLocalMatrix() {
        localMatrix.identity();
        localMatrix.translate(position);
        localMatrix.rotate(rotation);
        localMatrix.scale(scale);
    }

    /**
     * Update world model matrix.
     */
    public void updateWorldModelMatrix() {

        // parentMatrix * worldModelMatrix
        Entity parent = getParent();
        if (parent != null) {
            Matrix4f parentMatrix = parent.getWorldModelMatrix();
            parentMatrix.mul(localMatrix, worldModelMatrix);
        } else {

            // worldModelMatrix = localMatrix
            worldModelMatrix.set(localMatrix);
        }
    }

    /**
     * Get world model matrix.
     *
     * @return World model matrix.
     */
    @NonNull
    public Matrix4f getWorldModelMatrix() {
        return worldModelMatrix;
    }

    /**
     * Add {@link Component}. Note that only one Component can be added per class.
     *
     * @param component Component
     * @return {@code true} if Successfully added. Otherwise {@code false}.
     */
    public boolean add(@NonNull Component component) {
        final Class<? extends Component> componentClass = component.getClass();
        if (!components.containsKey(componentClass)) {
            component.setEntity(this);
            component.onAttach(this);
            components.put(componentClass, component);

            if (component.getClass().equals(GeometryComponent.class) || component.getClass().equals(SurfaceRendererComponent.class)) {
                this.renderable = hasComponent(GeometryComponent.class) && hasComponent(SurfaceRendererComponent.class);
            }

            return true;
        }
        return false;
    }

    /**
     * Remove {@link Component}.
     *
     * @param component Component.
     * @return {@code true} if Successfully removed. Otherwise {@code false}.
     */
    public boolean remove(@NonNull Component component) {
        return remove(component.getClass());
    }

    /**
     * Remove {@link Component} with associated class.
     *
     * @param clazz Class of Component.
     * @param <T>   Type
     * @return {@code true} if Successfully removed. Otherwise {@code false}.
     */
    public <T extends Component> boolean remove(@NonNull Class<T> clazz) {
        if (components.containsKey(clazz)) {
            Component component = components.get(clazz);
            component.onDetach(this);
            component.setEntity(null);
            components.remove(clazz);

            if (clazz.equals(SurfaceRendererComponent.class) || clazz.equals(GeometryComponent.class)) {
                this.renderable = hasComponent(GeometryComponent.class) && hasComponent(SurfaceRendererComponent.class);
            }

            return true;
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    public <T extends Component> T getComponent(@NonNull Class<T> clazz) {
        return (T) components.get(clazz);
    }

    public boolean hasComponent(@NonNull Class<? extends Component> clazz) {
        return components.containsKey(clazz);
    }

    /**
     * Add child {@link Entity}.
     *
     * @param child Child entity.
     * @return {@code true} if Successfully added. Otherwise {@code false}.
     */
    public boolean add(@NonNull Entity child) {
        final boolean added = children.add(child);
        if (added) {
            child.parent = this;
            child.setApp(app);
        }
        return added;
    }

    /**
     * Remove child {@link Entity}.
     *
     * @param child Child entity.
     * @return {@code true} if Successfully removed. Otherwise {@code false}.
     */
    public boolean remove(@NonNull Entity child) {
        final boolean removed = children.remove(child);
        if (removed) {
            child.parent = null;
        }
        return removed;
    }

    /**
     * Remove this from scene graph.
     *
     * @return {@code true} if Successfully removed. Otherwise {@code false}.
     * @since 3.0.17
     */
    public boolean remove() {
        return parent != null && parent.remove(this);
    }

    /**
     * Get children of Entity. Returned {@link List} can not be modified.
     *
     * @return Children of Entity.
     */
    @NonNull
    public List<Entity> getChildren() {
        return Collections.unmodifiableList(children);
    }

    /**
     * Get parent {@link Entity}.
     *
     * @return Parent entity.
     */
    @Nullable
    public Entity getParent() {
        return parent;
    }

    /**
     * For internal use only.
     *
     * @return native pointer value.
     */
    public long getNativePointer() {
        return nativePointer.get();
    }

    private void invalidateWorldModelMatrix() {

        worldMatrixUpdateRequired = true;

        // Propagate to children
        for (Entity child : children) {
            child.invalidateWorldModelMatrix();
        }
    }

    /**
     * Set Entity's local position.
     *
     * @param x X component of position
     * @param y Y component of position
     * @param z Z component of position
     */
    public void setPosition(float x, float y, float z) {
        this.position.set(x, y, z);
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local position.
     *
     * @param x X component of position
     */
    public void setX(float x) {
        this.position.x = x;
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local position.
     *
     * @param y Y component of position
     */
    public void setY(float y) {
        this.position.y = y;
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local position.
     *
     * @param z Z component of position
     */
    public void setZ(float z) {
        this.position.z = z;
        localMatrixUpdateRequired = true;
    }

    /**
     * Get Entity's local position.
     *
     * @return position
     */
    @NonNull
    public Vector3f getPosition() {
        return position;
    }

    /**
     * Set Entity's local position.
     *
     * @param position position
     */
    public void setPosition(@NonNull Vector3f position) {
        this.position.set(position);
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local scale.
     *
     * @param x X component of scale
     * @param y Y component of scale
     * @param z Z component of scale
     */
    public void setScale(float x, float y, float z) {
        this.scale.set(x, y, z);
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local scale.
     *
     * @param x X component of scale
     */
    public void setScaleX(float x) {
        this.scale.x = x;
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local scale.
     *
     * @param y Y component of scale
     */
    public void setScaleY(float y) {
        this.scale.y = y;
        localMatrixUpdateRequired = true;
    }

    /**
     * Set Entity's local scale.
     *
     * @param z Z component of scale
     */
    public void setScaleZ(float z) {
        this.scale.z = z;
        localMatrixUpdateRequired = true;
    }

    /**
     * Get Entity's local scale.
     *
     * @return scale
     */
    @NonNull
    public Vector3f getScale() {
        return scale;
    }

    /**
     * Set Entity's local scale.
     *
     * @param scale scale
     */
    public void setScale(@NonNull Vector3f scale) {
        this.scale.set(scale);
        localMatrixUpdateRequired = true;
    }

    /**
     * Get Entity's local rotation.
     *
     * @return rotation
     */
    @NonNull
    public Quaternionf getRotation() {
        return rotation;
    }

    /**
     * Set rotation, as a quaternion. Sets the transform's current rotation in quaternion terms.
     *
     * @param rotation rotation
     */
    public void setRotation(@NonNull Quaternionf rotation) {
        this.rotation.set(rotation);
        localMatrixUpdateRequired = true;
    }

    /**
     * Get {@link View}. This works only for Entity which is created by {@link #from(View)}
     * or has {@link SurfaceRendererComponent} which is
     * created by {@link SurfaceRendererComponent#from(View)}.
     *
     * @return View
     */
    @Nullable
    public View view() {
        SurfaceRendererComponent surfaceRendererComponent = getComponent(SurfaceRendererComponent.class);
        if (surfaceRendererComponent == null) {
            return null;
        }

        SurfaceRendererComponent.CanvasRenderer canvasRenderer = surfaceRendererComponent.getCanvasRenderer();
        if (canvasRenderer == null) {
            return null;
        }

        if (canvasRenderer instanceof SurfaceRendererComponent.ViewRenderer) {
            return ((SurfaceRendererComponent.ViewRenderer) canvasRenderer).getView();
        }

        return null;
    }

    /**
     * Short hand for {@code new EntityAnimator(entity)}.
     *
     * @return new EntityAnimator
     */
    @NonNull
    public EntityAnimator animate() {
        return new EntityAnimator(this);
    }

    private void updateOpacity() {

        SurfaceRendererComponent surfaceRendererComponent = getComponent(SurfaceRendererComponent.class);
        if (surfaceRendererComponent != null) {
            surfaceRendererComponent.setOpacity(getRenderingOpacity());
        }

        for (Entity child : children) {
            child.updateOpacity();
        }
    }

    private float parentOpacity() {
        return parent != null ? parent.getOpacity() * parent.parentOpacity() : 1.0f;
    }

    /**
     * Get actual opacity used in rendering.
     * This value can be different with value returned from {@link #getOpacity()}.
     *
     * @return Rendering opacity
     */
    public float getRenderingOpacity() {
        return opacity * parentOpacity();
    }

    /**
     * Get opacity set with {@link #setOpacity(float)}.
     * Actual opacity used in rendering can get with {@link #getRenderingOpacity()}.
     *
     * @return opacity
     */
    public float getOpacity() {
        return opacity;
    }

    /**
     * Set Entity's opacity.
     *
     * @param opacity Float in the range of 0.0 - 1.0 indicating how transparent the Entity is.
     */
    public void setOpacity(float opacity) {

        // Prevent out of range
        if (opacity < 0) {
            opacity = 0;
        } else if (opacity > 1) {
            opacity = 1;
        }

        // Do nothing if previous value is same to new value
        if (this.opacity == opacity) return;

        this.opacity = opacity;
        updateOpacityRequired = true;
    }

    /**
     * Get visibility of this {@link Entity}. Note that this will return <b>its own visibility</b>.
     *
     * @return visibility
     */
    public boolean isVisible() {
        return visible;
    }

    /**
     * Set visibility. If set to {@code false}, its children also become not be rendered.
     * Default value is {@code true}.
     *
     * @param visible visibility
     */
    public void setVisible(boolean visible) {
        this.visible = visible;
    }

    /**
     * @return {@code true} if this entity and ancestors are all visible. Otherwise {@code false}.
     */
    public boolean isShown() {

        if (!visible) return false;

        // Check parent visibility
        Entity parent = getParent();
        if (parent == null) return true;

        return parent.isShown();
    }

    public boolean isRenderable() {
        return renderable;
    }

    /**
     * Returns the component of type in the {@link Entity} or any of its children using depth first search.
     *
     * @param type Component class
     * @return Found {@link Component}. If no one is found, null.
     */
    public <T extends Component> T getComponentInChildren(@NonNull Class<T> type) {

        for (Entity child : children) {
            T component = child.getComponentInChildren(type);
            if (component != null) {
                return component;
            }
        }

        return getComponent(type);
    }

    /**
     * Returns the component of type in the {@link Entity} or any of its parents.
     * Recurses upwards until it finds a valid component. Returns null if no component found.
     *
     * @param type Component class
     * @return Found {@link Component}. If no one is found, null.
     */
    public <T extends Component> T getComponentInParent(@NonNull Class<T> type) {

        if (hasComponent(type)) {
            return getComponent(type);
        }

        return parent != null ? parent.getComponentInParent(type) : null;
    }
}

File: library/src/main/java/org/meganekkovr/Entity.kt
package org.meganekkovr

import android.graphics.drawable.Drawable
import android.view.View
import org.joml.Matrix4f
import org.joml.Quaternionf
import org.joml.Vector3f
import org.meganekkovr.animation.EntityAnimator

/**
 * Entity is a basic class in Meganekko. It is something on VR scene.
 * Entity has transform information such as [position][.setPosition],
 * [scale][.setScale], and [rotation][.setRotation].
 * It can have child Entities and [Component]s.
 * Entity which will be rendered has to have [GeometryComponent] and [SurfaceRendererComponent].
 */
open class Entity {
    private val _nativePointer = NativePointer.getInstance(newInstance())

    /**
     * For internal use only.
     *
     * @return native pointer value.
     */
    val nativePointer: Long
        get() {
            return _nativePointer.get()
        }

    private val components = mutableMapOf<Class<out Component>, Component>()

    /**
     * Get children of Entity.
     *
     * @return Children of Entity.
     */
    val children = mutableListOf<Entity>()

    /**
     * Get/Set Entity's local position.
     */
    var position = Vector3f()
        set(position) {
            field.set(position)
            localMatrixUpdateRequired = true
        }

    /**
     * Get/Set Entity's local scale.
     */
    var scale = Vector3f(1f, 1f, 1f)
        set(scale) {
            field.set(scale)
            localMatrixUpdateRequired = true
        }

    /**
     * Get/Set Entity's local rotation.
     */
    var rotation = Quaternionf()
        set(rotation) {
            field.set(rotation)
            localMatrixUpdateRequired = true
        }

    private val localMatrix = Matrix4f()

    /**
     * Get world model matrix.
     *
     * @return World model matrix.
     */
    val worldModelMatrix = Matrix4f()

    private val matrixValues = FloatArray(16)

    /**
     * This method is not valid until this is attached to [Scene].
     */
    // Propagate to children
    var app: MeganekkoApp? = null
        set(app) {
            field = app
            children.forEach { child -> child.app = app }
        }

    /**
     * Parent [Entity].
     */
    var parent: Entity? = null
        private set

    private var localMatrixUpdateRequired = true
    private var worldMatrixUpdateRequired = true

    var id: Int = 0

    /**
     * Get/Set opacity. Actual opacity used in rendering can get with [.getRenderingOpacity].
     */
    var opacity = 1.0f
        set(opacity) {
            // Prevent out of range
            val actualOpacity = if (opacity < 0) {
                0f
            } else if (opacity > 1) {
                1f
            } else {
                opacity
            }

            // Do nothing if previous value is same to new value
            if (this.opacity != actualOpacity) {
                field = actualOpacity
                updateOpacityRequired = true
            }
        }

    private var updateOpacityRequired: Boolean = false

    /**
     * Get/Set visibility of this [Entity]. Note that this will return **its own visibility**.
     * If set to `false`, its children also become not be rendered.
     * Default value is `true`.
     */
    var isVisible = true

    var isRenderable: Boolean = false
        private set // true if this has Geometry and Surface.

    /**
     * Get actual opacity used in rendering.
     * This value can be different with value returned from [.getOpacity].
     *
     * @return Rendering opacity
     */
    val renderingOpacity: Float
        get() = this.opacity * parentOpacity()

    /**
     * @return `true` if this entity and ancestors are all visible. Otherwise `false`.
     */
    // Check parent visibility
    val isShown: Boolean
        get() {

            if (!isVisible) return false

            return parent?.isShown ?: true
        }

    /**
     * Override this to create own native instance.
     * Native class must be derived from `mgn::Entity`.
     *
     * @return Native pointer value return from C++ `new`.
     */
    private external fun newInstance(): Long

    /**
     * Set string id.
     *
     * @param id ID
     */
    fun setId(id: String) {
        this.id = id.hashCode()
    }

    /**
     * Find [Entity] from children. Typically, id is `R.id.xxx` value.
     *
     * @param id ID
     * @return Found Entity or `null` if it has no matched Entity with id.
     */
    fun findById(id: Int): Entity? {

        if (this.id == id) return this

        children.forEach { child ->
            val found = child.findById(id)
            if (found != null) return found
        }

        return null
    }

    /**
     * Find [Entity] from children.
     *
     * @param id ID
     * @return Found Entity or `null` if it has no matched Entity with id.
     */
    fun findById(id: String): Entity? {
        return findById(id.hashCode())
    }

    /**
     * Called at every frame update.
     *
     * @param frame Frame information
     */
    open fun update(frame: FrameInput) {

        // Notify to components
        components.values.forEach { it.update(frame) }

        // Notify to children
        children.forEach { it.update(frame) }

        // Update local model matrix if necessary.
        if (localMatrixUpdateRequired) {
            updateLocalMatrix()
            invalidateWorldModelMatrix()
            localMatrixUpdateRequired = false
        }

        // Update world model matrix if necessary.
        if (worldMatrixUpdateRequired) {
            updateWorldModelMatrix()
            worldMatrixUpdateRequired = false

            // Update native side values
            worldModelMatrix.get(matrixValues)
            setWorldModelMatrix(nativePointer, matrixValues)
        }

        // Update opacity if necessary.
        if (updateOpacityRequired) {
            updateOpacity()
            updateOpacityRequired = false
        }
    }

    /**
     * Update local matrix.
     */
    private fun updateLocalMatrix() {
        localMatrix.identity()
        localMatrix.translate(this.position)
        localMatrix.rotate(this.rotation)
        localMatrix.scale(this.scale)
    }

    /**
     * Update world model matrix.
     */
    private fun updateWorldModelMatrix() {

        // parentMatrix * worldModelMatrix
        val parent = parent
        if (parent != null) {
            val parentMatrix = parent.worldModelMatrix
            parentMatrix.mul(localMatrix, worldModelMatrix)
        } else {

            // worldModelMatrix = localMatrix
            worldModelMatrix.set(localMatrix)
        }
    }

    /**
     * Add [Component]. Note that only one Component can be added per class.
     *
     * @param component Component
     * @return `true` if Successfully added. Otherwise `false`.
     */
    fun add(component: Component): Boolean {
        val componentClass = component.javaClass
        if (!components.containsKey(componentClass)) {
            component.entity = this
            component.onAttach(this)
            components[componentClass] = component

            if (component.javaClass == GeometryComponent::class.java || component.javaClass == SurfaceRendererComponent::class.java) {
                this.isRenderable = hasComponent<GeometryComponent>() && hasComponent<SurfaceRendererComponent>()
            }

            return true
        }
        return false
    }

    /**
     * Remove [Component].
     *
     * @param component Component.
     * @return `true` if Successfully removed. Otherwise `false`.
     */
    fun remove(component: Component): Boolean {
        return remove(component.javaClass)
    }

    /**
     * Remove [Component] with associated class.
     *
     * @param clazz Class of Component.
     * @param <T>   Type
     * @return `true` if Successfully removed. Otherwise `false`.
     */
    fun <T : Component> remove(clazz: Class<T>): Boolean {
        if (components.containsKey(clazz)) {
            val component = components[clazz]
            component!!.onDetach(this)
            component.entity = null
            components.remove(clazz)

            if (clazz == SurfaceRendererComponent::class.java || clazz == GeometryComponent::class.java) {
                this.isRenderable = hasComponent<GeometryComponent>() && hasComponent<SurfaceRendererComponent>()
            }

            return true
        }
        return false
    }

    inline fun <reified T : Component> getComponent(): T? {
        return getComponent(T::class.java)
    }

    fun <T : Component> getComponent(clazz: Class<T>): T? {
        return components[clazz] as T?
    }

    inline fun <reified T : Component> hasComponent(): Boolean {
        return hasComponent(T::class.java)
    }

    fun hasComponent(clazz: Class<out Component>): Boolean {
        return components.containsKey(clazz)
    }

    /**
     * Add child [Entity].
     *
     * @param child Child entity.
     * @return `true` if Successfully added. Otherwise `false`.
     */
    fun add(child: Entity): Boolean {
        val added = children.add(child)
        if (added) {
            child.parent = this
            child.app = this.app
        }
        return added
    }

    /**
     * Remove child [Entity].
     *
     * @param child Child entity.
     * @return `true` if Successfully removed. Otherwise `false`.
     */
    fun remove(child: Entity): Boolean {
        val removed = children.remove(child)
        if (removed) {
            child.parent = null
        }
        return removed
    }

    /**
     * Remove this from scene graph.
     *
     * @return `true` if Successfully removed. Otherwise `false`.
     * @since 3.0.17
     */
    fun remove(): Boolean {
        return parent?.remove(this) ?: false
    }

    private fun invalidateWorldModelMatrix() {

        worldMatrixUpdateRequired = true

        // Propagate to children
        children.forEach(Entity::invalidateWorldModelMatrix)
    }

    /**
     * Set Entity's local position.
     *
     * @param x X component of position
     * @param y Y component of position
     * @param z Z component of position
     */
    fun setPosition(x: Float, y: Float, z: Float) {
        this.position.set(x, y, z)
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local position.
     *
     * @param x X component of position
     */
    fun setX(x: Float) {
        this.position.x = x
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local position.
     *
     * @param y Y component of position
     */
    fun setY(y: Float) {
        this.position.y = y
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local position.
     *
     * @param z Z component of position
     */
    fun setZ(z: Float) {
        this.position.z = z
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local scale.
     *
     * @param x X component of scale
     * @param y Y component of scale
     * @param z Z component of scale
     */
    fun setScale(x: Float, y: Float, z: Float) {
        this.scale.set(x, y, z)
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local scale.
     *
     * @param x X component of scale
     */
    fun setScaleX(x: Float) {
        this.scale.x = x
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local scale.
     *
     * @param y Y component of scale
     */
    fun setScaleY(y: Float) {
        this.scale.y = y
        localMatrixUpdateRequired = true
    }

    /**
     * Set Entity's local scale.
     *
     * @param z Z component of scale
     */
    fun setScaleZ(z: Float) {
        this.scale.z = z
        localMatrixUpdateRequired = true
    }

    /**
     * Get [View]. This works only for Entity which is created by [.from]
     * or has [SurfaceRendererComponent] which is
     * created by [SurfaceRendererComponent.from].
     *
     * @return View
     */
    fun view(): View? {
        val surfaceRendererComponent = getComponent<SurfaceRendererComponent>() ?: return null

        val canvasRenderer = surfaceRendererComponent.canvasRenderer ?: return null

        return (canvasRenderer as? SurfaceRendererComponent.ViewRenderer)?.view

    }

    /**
     * Short hand for `new EntityAnimator(entity)`.
     *
     * @return new EntityAnimator
     */
    fun animate(): EntityAnimator {
        return EntityAnimator(this)
    }

    private fun updateOpacity() {

        val surfaceRendererComponent = getComponent<SurfaceRendererComponent>()
        surfaceRendererComponent?.opacity = renderingOpacity

        children.forEach(Entity::updateOpacity)
    }

    private fun parentOpacity(): Float {
        return if (parent != null) parent!!.opacity * parent!!.parentOpacity() else 1.0f
    }

    /**
     * Returns the component of type in the [Entity] or any of its children using depth first search.
     *
     * @param type Component class
     * @return Found [Component]. If no one is found, null.
     */
    fun <T : Component> getComponentInChildren(type: Class<T>): T? {

        children.forEach { child ->
            val component = child.getComponentInChildren(type)
            if (component != null) {
                return component
            }
        }

        return getComponent(type)
    }

    /**
     * Returns the component of type in the [Entity] or any of its parents.
     * Recurses upwards until it finds a valid component. Returns null if no component found.
     *
     * @param type Component class
     * @return Found [Component]. If no one is found, null.
     */
    fun <T : Component> getComponentInParent(type: Class<T>): T? {

        return if (hasComponent(type)) {
            getComponent(type)
        } else {
            parent?.getComponentInParent(type)
        }
    }

    private external fun setWorldModelMatrix(nativePtr: Long, matrix: FloatArray)

    companion object {

        /**
         * Create Entity from [View]. New Entity has plane geometry.
         *
         * @param view View for surface.
         * @return new Entity
         */
        fun from(view: View): Entity {

            return Entity().apply {
                add(SurfaceRendererComponent.from(view))
                add(GeometryComponent.from(view))
            }
        }

        /**
         * Create Entity from [Drawable]. New Entity has plane geometry.
         *
         * @param drawable Drawable for surface.
         * @return new Entity
         */
        fun from(drawable: Drawable): Entity {

            return Entity().apply {
                add(SurfaceRendererComponent.from(drawable))
                add(GeometryComponent.from(drawable))
            }
        }
    }
}
